libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/RoundTrip.hs

```diff
@@ -0,0 +1,197 @@
+roundTripEraTypeExpectation ::
+  forall era t.
+  (Era era, Show (t era), Eq (t era), EncCBOR (t era), DecCBOR (t era)) =>
+  t era ->
+  Expectation
+roundTripEraTypeExpectation = roundTripEraExpectation @era @(t era)
+
+-- | QuickCheck property spec that uses `roundTripAnnEraTypeExpectation`
+roundTripAnnEraTypeSpec ::
+  forall era t.
+  ( Era era
+  , Show (t era)
+  , Eq (t era)
+  , ToCBOR (t era)
+  , DecCBOR (Annotator (t era))
+  , Arbitrary (t era)
+  ) =>
+  Spec
+roundTripAnnEraTypeSpec =
+  prop (show (typeRep $ Proxy @(t era))) $ roundTripAnnEraTypeExpectation @era @t
+
+-- | Same as `roundTripAnnEraExpectation`, but is not suitable for type families.
+roundTripAnnEraTypeExpectation ::
+  forall era t.
+  ( Era era
+  , Show (t era)
+  , Eq (t era)
+  , ToCBOR (t era)
+  , DecCBOR (Annotator (t era))
+  ) =>
+  t era ->
+  Expectation
+roundTripAnnEraTypeExpectation = roundTripAnnEraExpectation @era @(t era)
+
+-- | QuickCheck property spec that uses `roundTripShareEraExpectation`
+roundTripShareEraSpec ::
+  forall era t.
+  (Era era, Show t, Eq t, EncCBOR t, DecShareCBOR t, Arbitrary t) =>
+  Spec
+roundTripShareEraSpec =
+  prop (show (typeRep $ Proxy @t)) $ roundTripShareEraExpectation @era @t
+
+-- | Roundtrip CBOR testing for types and type families that implement
+-- EncCBOR/DecShareCBOR. Requires TypeApplication of an @@era@
+roundTripShareEraExpectation ::
+  forall era t.
+  (Era era, Show t, Eq t, EncCBOR t, DecShareCBOR t) =>
+  t ->
+  Expectation
+roundTripShareEraExpectation =
+  roundTripRangeExpectation
+    (mkTrip encCBOR decNoShareCBOR)
+    (eraProtVerLow @era)
+    (eraProtVerHigh @era)
+
+-- | QuickCheck property spec that uses `roundTripShareEraTypeExpectation`
+roundTripShareEraTypeSpec ::
+  forall era t.
+  (Era era, Show (t era), Eq (t era), EncCBOR (t era), DecShareCBOR (t era), Arbitrary (t era)) =>
+  Spec
+roundTripShareEraTypeSpec =
+  prop (show (typeRep $ Proxy @(t era))) $ roundTripShareEraTypeExpectation @era @t
+
+-- | Roundtrip CBOR testing for types that implement EncCBOR/DecShareCBOR. Unlike
+-- `roundTripShareEraExpectation`, this function can't be used with type families, but the
+-- types of this function are unambiguous.
+roundTripShareEraTypeExpectation ::
+  forall era t.
+  (Era era, Show (t era), Eq (t era), EncCBOR (t era), DecShareCBOR (t era)) =>
+  t era ->
+  Expectation
+roundTripShareEraTypeExpectation = roundTripShareEraExpectation @era @(t era)
+
+-- | CBOR RoundTrip spec for all the core types and type families that are parametrized on era.
+roundTripCoreEraTypesSpec ::
+  forall era.
+  ( EraTx era
+  , Arbitrary (Tx era)
+  , Arbitrary (TxBody era)
+  , Arbitrary (TxOut era)
+  , Arbitrary (TxCert era)
+  , Arbitrary (TxWits era)
+  , Arbitrary (TxAuxData era)
+  , Arbitrary (Value era)
+  , Arbitrary (CompactForm (Value era))
+  , Arbitrary (Script era)
+  , Arbitrary (PParams era)
+  , Arbitrary (PParamsUpdate era)
+  ) =>
+  Spec
+roundTripCoreEraTypesSpec = do
```

Comment:
Is this function meant to be invoked in each era (in the respective RoundTrip test) ?

Reply:
Yes. And it is

Reply:
Ah yes! Via roundTripShelleyCommonSpec and roundTripAlonzoCommonSpec. Sorry, I missed that :+1: